notes.md

```diff
@@ -0,0 +1,264 @@
+to another process so as not to hurt read performance times for posts.
+The user achieves this by first implementing a listener for the caching:
+
+Note: This example is for illustrative purposes only. There are of course
+many other ways to achieve this.
+
+
+### Declaring a Listener
+First, define the listener in their app's `tests.listeners.py` file:
+
+```python
+post_reads_per_date_cache = defaultdict(dict)  # dummy cache for illustrative purposes only
+
+@listen('post_reads_per_date')
+def post_reads_per_date(
+        *,
+        post_id: int,
+        date: datetime.date,
+):
+    print(f'Processing post_reads_per_date with args {post_id}, {date}')
+    print(f'Cache before: {post_reads_per_date_cache}')
+    current_count = post_reads_per_date_cache[date].get(post_id, 0)
+    post_reads_per_date_cache[date][post_id] = current_count + 1
+    print(f'Cache after: {post_reads_per_date_cache}')
+```
+
+The user must also ensure that this `listeners` module is imported into
+that app's config class (similar to signals):
+```python
+class TestsConfig(AppConfig):
+    name = 'tests'
+
+    def ready(self):
+        import tests.listeners
+```
+
+#### Registering listeners
+When the project is loaded, the `listen` decorator is called. This
+executes this piece of logic:
+```python
+CustomPayloadChannel.register(callback, channel_name)
+```
+This creates an instance of CustomPayloadChannel and adds it to a global
+registry of "channels".
+
+##### Channels
+As in the postgres listen/notify protocol, messages are sent and received
+via named channels. The objects defined in channels.py are python abstractions
+of these channels. Instances of these classes have the following responsibilities:
+
+- Define an appropriate channel name if one has not been supplied.
+    - Undecided on how to best generate the name - see comments in _create_name.
+    - This is why in the examples I just define the names manually.
+- Listen to notifications on the channel matching their name.
+    - See the `listen` method
+- Deserialize notifications appropriately for the stored callback (the one
+  appearing under the `listen` decorator).
+    - I would really like the user to be able to notify the above channel with
+      the same payload as is accepted by the `post_reads_per_date` callback.
+    - I also want the callback to be able to accept a reasonable variety of different
+      python objects.
+    - So e.g. like `notify('post_reads_per_date', post_id=pk, date=datetime.date.today())`
+    - This means we need a way to easily serialize and then deserialize these arguments
+      back into python objects.
+    - The obvious choice for this is would be to use `pickle` as the serialization method. This
+      however did not work easily as the byte string pickling produced was not accepted by the
+      `pg_notify` function. Some sort of replace on the bad characters might let it, but that
+      felt a bit hacky.
```

Comment:
I don't know LISTEN/NOTIFY in depth, but i think you are going to get a long ways just serializing to JSON and using regular JSON objects. I'm not sure how well pickle objects are going to work being sent from a trigger as well.

In pghistory, it is taking a dictionary object from a user, JSON serializing it, and then loading it as a postgres variable here - https://github.com/Opus10/django-pghistory/blob/master/pghistory/tracking.py#L40

When that variable is loaded, I then have a postgres function that just casts it into JSONB https://github.com/Opus10/django-pghistory/blob/master/pghistory/models.py#L55

I guess what I'm getting at is that I think standard JSON is going to work quite well, and it's really easy just to pass a JSON string to posgres. I'm assuming it will work fine for listen/notify. The only drawback is that the user needs to pass serializable data to listen/notify instead of Python objects.

Reply:
Nevermind, I just read a few lines later and see you are using json!

Reply:
Yeah - using purely JSON serialize/deserialize in trigger based notify. For the manual notify, the annotations of the listeners allow us to deserialize back to the original python variable type (so e.g. if we notify with a datetime.date object, we can serialize to JSON and then deserialize back to datetime.date in the listener callback).  Right now I support python iterables of primitive types in regards to this. Eventually I'd like to be able to support django model instances too so we can write listen callbacks more like regular signals (have something kind of like that for trigger based listeners with TriggerPayload).

Maybe it's not a big deal, but a few times when using celery tasks I've found it really annoying to not be able just to use things like datetime.dates and have to remember to serialize/deserialize them myself.